src/wayland/compositor/transaction.rs

```diff
@@ -108,6 +130,114 @@ impl Blocker for Barrier {
+    pub fn with_surfaces<'a>(ping: Ping, surfaces: impl IntoIterator<Item = &'a WlSurface>) -> Self {
+        let barrier = Barrier::new(false);
+        let ready = Arc::new(AtomicBool::new(false));
+        let barrier = Self {
+            barrier,
+            ping,
+            ready,
+            surfaces: Arc::new(Mutex::new(HashMap::new())),
+        };
+        barrier.register_surfaces(surfaces);
+        barrier
+    }
+
+    /// Register surfaces to be tracked by this barrier.
+    ///
+    /// This will automatically insert a blocker that will resolve on [`SurfaceBarrier::release`].
+    pub fn register_surfaces<'a>(&self, surfaces: impl IntoIterator<Item = &'a WlSurface>) {
+        let mut lock = self.surfaces.lock().unwrap();
+        for surface in surfaces {
+            if let std::collections::hash_map::Entry::Vacant(vacant_entry) = lock.entry(surface.downgrade()) {
+                vacant_entry.insert(false);
+                add_blocker(
+                    surface,
+                    SurfaceBarrierBlocker {
+                        barrier: self.barrier.clone(),
+                        ping: self.ping.clone(),
+                        ready: self.ready.clone(),
+                        surfaces: self.surfaces.clone(),
+                    },
+                );
+            }
+        }
+        self.ready.store(false, std::sync::atomic::Ordering::Release);
+    }
+
+    /// Release this barrier and clear the blocker on all tracked surfaces
+    pub fn release<D: CompositorHandler + 'static>(&self, dh: &DisplayHandle, state: &mut D) {
+        self.barrier.signal();
+        #[allow(clippy::mutable_key_type)]
+        let surfaces = { std::mem::take(&mut *self.surfaces.lock().unwrap()) };
+        for (surface, _) in surfaces {
+            let Ok(surface) = surface.upgrade() else {
+                continue;
+            };
+            let Some(client) = surface.client() else {
+                continue;
+            };
+            state.client_compositor_state(&client).blocker_cleared(state, dh);
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct SurfaceBarrierBlocker {
+    barrier: Barrier,
+    ping: Ping,
+    ready: Arc<AtomicBool>,
+    surfaces: Arc<Mutex<HashMap<Weak<WlSurface>, bool>>>,
+}
+
+impl Blocker for SurfaceBarrierBlocker {
+    fn state(&self) -> BlockerState {
+        self.barrier.state()
+    }
+
+    fn kind(&self) -> BlockerKind {
+        BlockerKind::Delayed
+    }
+
+    fn notify(&self, surface: &WlSurface) {
+        let mut surfaces = self.surfaces.lock().unwrap();
+        surfaces
+            .entry(surface.downgrade())
+            .and_modify(|state| *state = true);
+        if surfaces
+            .iter()
+            .all(|(surface, state)| *state || !surface.is_alive())
+        {
+            let signaled = self.ready.swap(true, std::sync::atomic::Ordering::Acquire);
+            if !signaled {
+                self.ping.ping();
```

Comment:
IIRC in niri I didn't use the Ping source for the actual ready notification because it would arrive only on the next event loop iteration. I'm not sure if this is a problem here.

Reply:
Interesting. I am also not sure this is a problem in this case. We could use a channel instead maybe?